IdnoPlugins/IndiePub/Pages/MicroPub/Endpoint.php

```diff
@@ -227,20 +231,58 @@ function post()
                             exit;
                         } else {
                             \Idno\Core\Idno::site()->triggerEvent('indiepub/post/failure', ['page' => $this]);
-                            $this->setResponse(500);
-                            echo "Couldn't create {$type}";
-                            exit;
+                            $this->error(
+                                400, 'invalid_request',
+                                "Couldn't create {$type}"
+                            );
                         }
 
                     }
 
                 } else {
                     \Idno\Core\Idno::site()->triggerEvent('indiepub/post/failure', ['page' => $this]);
-                    $this->setResponse(500);
-                    echo "Couldn't find content type {$type}";
-                    exit;
+                    $this->error(
+                        400, 'invalid_request',
+                        "Couldn't find content type {$type}"
+                    );
+                }
+            }
 
+            /**
+             * Reply with a micropub error response and exit.
+             *
+             * @param int    $statusCode  HTTP status code
+             * @param string $error       Micropub error code
+             * @param string $description Human-readable error description
+             *
+             * @return void
+             */
+            protected function error($statusCode, $error, $description)
+            {
+                $msgs = \Idno\Core\Idno::site()->session()->getMessages();
+                foreach ($msgs as $msg) {
+                    if ($msg['message_type'] == 'alert-danger') {
+                        $description .= ': ' . $msg['message'];
+                    }
                 }
+
+                $this->setResponse($statusCode);
+                header('Content-type: application/json');
+                echo json_encode(
+                    array(
+                        'error'             => $error,
+                        'error_description' => $description,
+                    )
+                ) . "\n";
```

Comment:
I tried that; it works but adds the `messages` key - unconditionally. This isn't against the micropub spec since that one does not have a schema for errors anyway.